src/core/net/ip6_mpl.cpp

```diff
@@ -85,79 +85,141 @@ void Mpl::InitOption(OptionMpl &aOption, const Address &aAddress)
+                curCount++;
+            }
 
-exit:
-    return error;
-}
+            if (maxCount < curCount)
+            {
+                // look to evict an entry from the seed with the most entries
+                evict    = &mSeedSet[i - curCount];
+                maxCount = curCount;
+            }
 
-void Mpl::UpdateBufferedSet(uint16_t aSeedId, uint8_t aSequence)
-{
-    int8_t                     diff;
-    MplBufferedMessageMetadata messageMetadata;
+            curCount = 0;
+        }
 
-    Message *message     = mBufferedMessageSet.GetHead();
-    Message *nextMessage = NULL;
+        if (aSeedId == mSeedSet[i].GetSeedId())
+        {
+            // have existing entries for aSeedId
 
-    // Check if multicast forwarding is enabled.
-    VerifyOrExit(GetTimerExpirations() > 0);
+            if (aSequence == mSeedSet[i].GetSequence())
+            {
+                // already received, drop message
+                ExitNow(error = OT_ERROR_DROP);
+            }
+            else if (insert == NULL && aSequence < mSeedSet[i].GetSequence())
+            {
+                // insert in order of sequence
+                insert = &mSeedSet[i];
+                curCount++;
+            }
+        }
+    }
 
-    while (message != NULL)
+    if (evict->GetLifetime() != 0)
     {
-        nextMessage = message->GetNext();
-        messageMetadata.ReadFrom(*message);
+        // no free entries available, look to evict an existing entry
 
-        if (messageMetadata.GetSeedId() == aSeedId)
+        if (aSeedId == mSeedSet[kNumSeedEntries - 1 - curCount].GetSeedId() && insert == NULL)
         {
-            diff = static_cast<int8_t>(aSequence - messageMetadata.GetSequence());
+            // insert at end of existing group
+            insert = &mSeedSet[kNumSeedEntries];
+            curCount++;
+        }
 
-            if (diff > 0)
-            {
-                // Stop retransmitting MPL Data Message that is consider to be old.
-                mBufferedMessageSet.Dequeue(*message);
-                message->Free();
-            }
+        if (maxCount < curCount)
+        {
+            // look to evict an entry from the seed with the most entries
+            evict    = &mSeedSet[kNumSeedEntries - 1 - curCount];
+            maxCount = curCount;
+        }
 
-            break;
+        // require evict group size to have >= 2 entries
+        VerifyOrExit(maxCount > 1, error = OT_ERROR_DROP);
+
+        if (insert == NULL)
+        {
+            // no existing entries for aSeedId
+            insert = evict;
+        }
+        else
+        {
+            // require require Sequence to be larger than oldest stored Sequence in group
+            VerifyOrExit(insert > mSeedSet && aSeedId == (insert - 1)->GetSeedId(), error = OT_ERROR_DROP);
         }
+    }
 
-        message = nextMessage;
+    if (evict > insert)
+    {
+        memmove(insert, insert + 1, static_cast<size_t>(evict - insert) * sizeof(MplSeedEntry));
```

Comment:
I didn't debug it very deeply however i think there is some unwanted memory access/modification here.

On our stress tests devices ended up in HardFault state. I was able to reproduce this error using CLI (with one modification):
```
-            VerifyOrExit(0 < value && value <= Timer::kMaxDt / 1000, error = OT_ERROR_INVALID_ARGS);
-            mInterval = static_cast<uint32_t>(value) * 1000;
+            VerifyOrExit(0 < value && value <= Timer::kMaxDt, error = OT_ERROR_INVALID_ARGS);
+            mInterval = static_cast<uint32_t>(value);
```

If you start sending Echo Requests with short interval e.g. 10ms between two routers: `ping ff03::1 10 100 10` the following HardFault appears:

```
HardFault_Handler () at /mnt/c/git/openthread/examples/../examples/platforms/nrf52840/../../../third_party/NordicSemiconductor/nrfx/mdk/gcc_startup_nrf52840.S:300
300     /mnt/c/git/openthread/examples/../examples/platforms/nrf52840/../../../third_party/NordicSemiconductor/nrfx/mdk/gcc_startup_nrf52840.S: No such file or directory.
(gdb) where
#0  HardFault_Handler () at /mnt/c/git/openthread/examples/../examples/platforms/nrf52840/../../../third_party/NordicSemiconductor/nrfx/mdk/gcc_startup_nrf52840.S:300
#1  <signal handler called>
#2  0x00027516 in memcmp (m1=<optimized out>, m2=<optimized out>, n=<optimized out>) at ../../../../../../../newlib/libc/string/memcmp.c:64
#3  0x0000a63e in ot::Ip6::Address::operator== (this=this@entry=0x2003ff38, aOther=...) at /mnt/c/git/openthread/examples/../src/core/net/ip6_address.cpp:246
#4  0x0000a914 in ot::Ip6::Mpl::InitOption (this=this@entry=0x20001488 <ot::sInstanceRaw+2192>, aOption=..., aAddress=...) at /mnt/c/git/openthread/examples/../src/core/net/ip6_mpl.cpp:74
#5  0x00009938 in ot::Ip6::Ip6::AddMplOption (this=this@entry=0x20001448 <ot::sInstanceRaw+2128>, aMessage=..., aHeader=...) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:205
#6  0x0000a0f8 in ot::Ip6::Ip6::SendDatagram (this=0x20001448 <ot::sInstanceRaw+2128>, aMessage=..., aMessageInfo=..., aIpProto=aIpProto@entry=ot::Ip6::kProtoIcmp6)
    at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:479
#7  0x00009582 in ot::Ip6::Icmp::SendEchoRequest (this=0x20001474 <ot::sInstanceRaw+2172>, aMessage=..., aMessageInfo=..., aIdentifier=<optimized out>) at /mnt/c/git/openthread/examples/../src/core/net/icmp6.cpp:99
#8  0x00002434 in ot::Cli::Interpreter::HandlePingTimer (this=0x200000e4 <ot::Cli::sCliUartRaw+4>) at /mnt/c/git/openthread/examples/../src/cli/cli.cpp:1977
#9  0x0001e5c2 in nrf5AlarmProcess (aInstance=aInstance@entry=0x20000bf8 <ot::sInstanceRaw>) at /mnt/c/git/openthread/examples/../examples/platforms/nrf52840/alarm.c:508
#10 0x0001f07e in otSysProcessDrivers (aInstance=aInstance@entry=0x20000bf8 <ot::sInstanceRaw>) at /mnt/c/git/openthread/examples/../examples/platforms/nrf52840/system.c:137
#11 0x0000034e in main (argc=0, argv=0x0 <_reclaim_reent>) at /mnt/c/git/openthread/examples/../examples/apps/cli/main.c:114
```

I couldn't reproduce this problem using previous commit on your branch.

Reply:
Thanks for testing this! I believe I've addressed the issue in my latest push. Specifically, changing `kNumSeedEntries - 1 - curCount` to `kNumSeedEntries - curCount`.

Reply:
Thanks for looking into it! If that is valuable, before merge I can run stress tests again tomorrow morning.

Reply:
If you're willing to test again, I can wait for the results :)

Reply:
Hello @jwhui 

Unfortunately fix didn't help, although i have enough GDB information to know what happened.

The HardFault from above suggested that the content of `mMatchingAddress` field in Mpl class is overwritten. This varaible is above `mSeedSet`. So i've added additional asserts to check where this unwanted access takes place:

```
     if (evict > insert)
     {
+        assert((uint32_t)insert > (uint32_t)&mMatchingAddress);
         memmove(insert, insert + 1, static_cast<size_t>(evict - insert) * sizeof(MplSeedEntry));
     }
     else if (evict < insert)
     {
+        assert((uint32_t)evict > (uint32_t)&mMatchingAddress);
         memmove(evict, evict + 1, static_cast<size_t>(insert - 1 - evict) * sizeof(MplSeedEntry));
         insert--;
     }
```

I found out that in stress tests, the second assert is triggered. The issues is related with `curCount` value overflow (> 32).

```
(gdb) where
#0  0x0001b320 in ot::Ip6::Mpl::UpdateSeedSet (this=0x20001488 <ot::sInstanceRaw+2192>, aSeedId=56320, aSequence=32 ' ') at /mnt/c/git/openthread/examples/../src/core/net/ip6_mpl.cpp:214
#1  0x0001b60e in ot::Ip6::Mpl::ProcessOption (this=0x20001488 <ot::sInstanceRaw+2192>, aMessage=..., aAddress=..., aIsOutbound=true) at /mnt/c/git/openthread/examples/../src/core/net/ip6_mpl.cpp:305
#2  0x000192a8 in ot::Ip6::Ip6::HandleOptions (this=0x20001448 <ot::sInstanceRaw+2128>, aMessage=..., aHeader=..., aForward=@0x2003fee8: true) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:581
#3  0x00019506 in ot::Ip6::Ip6::HandleExtensionHeaders (this=0x20001448 <ot::sInstanceRaw+2128>, aMessage=..., aHeader=..., aNextHeader=@0x2003ff59: 0 '\000', aForward=true, aReceive=false) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:645
#4  0x00019a06 in ot::Ip6::Ip6::HandleDatagram (this=0x20001448 <ot::sInstanceRaw+2128>, aMessage=..., aNetif=0x0 <_reclaim_reent>, aInterfaceId=1 '\001', aLinkMessageInfo=0x0 <_reclaim_reent>, aFromNcpHost=false) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:898
#5  0x00019144 in ot::Ip6::Ip6::HandleSendQueue (this=0x20001448 <ot::sInstanceRaw+2128>) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:545
#6  0x0001910a in ot::Ip6::Ip6::HandleSendQueue (aTasklet=...) at /mnt/c/git/openthread/examples/../src/core/net/ip6.cpp:535
#7  0x0000c3ec in ot::Tasklet::RunTask (this=0x20001468 <ot::sInstanceRaw+2160>) at /mnt/c/git/openthread/examples/../src/core/common/tasklet.hpp:93
#8  0x0000c56a in ot::TaskletScheduler::ProcessQueuedTasklets (this=0x20000c00 <ot::sInstanceRaw+8>) at /mnt/c/git/openthread/examples/../src/core/common/tasklet.cpp:112
#9  0x00007d62 in otTaskletsProcess (aInstance=0x20000bf8 <ot::sInstanceRaw>) at /mnt/c/git/openthread/examples/../src/core/api/tasklet_api.cpp:51
#10 0x0000035e in main (argc=0, argv=0x0 <_reclaim_reent>) at /mnt/c/git/openthread/examples/../examples/apps/cli/main.c:113
(gdb) p curCount
$1 = 33 '!'
```

I see that `curCount` may be double incremented in some cases. Hope that helps :)

Reply:
@LuDuda, thanks again for testing and digging into the issue!

I believe I've addressed the issue in the latest push. I've also verified that the asserts/hard faults no longer occur with the ping6 test you've described above. Thanks for keeping me honest!

Reply:
Hehe, the main motivation was just to avoid hard faults 😄 I verified again, seems to work like a charm 👍

Reply:
Thanks again, @LuDuda!